File: go-sdk/examples/two_hop_swap.go
// Package examples provides reusable Ifx business transaction planners.
package examples

import (
	"github.com/gagliardetto/solana-go"
	"github.com/ifx-run/ifx/go-sdk/scratch"
	"github.com/ifx-run/ifx/go-sdk/structuredcpi"
)

// TwoHopTokenSwapAccounts wire accounts for A→USDC→B same-tx orchestration.
type TwoHopTokenSwapAccounts struct {
	// UserUsdcAta is the intermediate mint ATA (e.g. USDC); must exist before this tx.
	UserUsdcAta solana.PublicKey
}

// TwoHopTokenSwapInstructions hop templates for two-hop swap.
type TwoHopTokenSwapInstructions struct {
	Hop1         solana.Instruction
	Hop2Template solana.Instruction
	Hop2Deliver  *solana.Instruction // optional deliver leg after patched hop2
}

// PlanTwoHopTokenSwapInstructions returns reset → hop1 → let USDC → patched hop2 [→ deliver].
// Mirrors sdk/examples/two-hop-token-swap.ts.
func PlanTwoHopTokenSwapInstructions(
	s *scratch.FrameScratch,
	accounts TwoHopTokenSwapAccounts,
	hops TwoHopTokenSwapInstructions,
) ([]solana.Instruction, error) {
	out := []solana.Instruction{s.IxReset(), hops.Hop1}

	b := s.LetBuilder()
	usdcOut, err := b.SplTokenAmount(accounts.UserUsdcAta)
	if err != nil {
		return nil, err
	}
	letIx, err := b.BuildIx()
	if err != nil {
		return nil, err
	}
	out = append(out, letIx)

	hop2Built, err := structuredcpi.StructuredCpi(
		hops.Hop2Template,
		structuredcpi.StructuredCpiPatch.TokenTransfer(structuredcpi.AsFrameValue(usdcOut)),
	)
	if err != nil {
		return nil, err
	}
	hop2, err := hop2Built.Build(nil)
	if err != nil {
		return nil, err
	}
	cpiIx, err := s.IxCpi(hop2)
	if err != nil {
		return nil, err
	}
	out = append(out, cpiIx)

	if hops.Hop2Deliver != nil {
		out = append(out, *hops.Hop2Deliver)
	}
	return out, nil
}

File: rust-sdk/tests/common/planners/two_hop_swap.rs
//! Two-hop SPL swap A → USDC → B — mirrors [`sdk/examples/two-hop-token-swap.ts`](../../../../sdk/examples/two-hop-token-swap.ts).

use ifx_sdk::patched_cpi::{build_structured_cpi, frame_value, structured_token_transfer};
use ifx_sdk::scratch::FrameScratch;
use ifx_sdk::ScratchError;
use solana_sdk::instruction::Instruction;
use solana_sdk::pubkey::Pubkey;

#[derive(Clone, Debug)]
pub struct TwoHopTokenSwapAccounts {
    /// Intermediate mint ATA (e.g. USDC); must exist before this tx.
    pub user_usdc_ata: Pubkey,
}

#[derive(Clone, Debug)]
pub struct TwoHopTokenSwapInstructions {
    /// Hop 1: A → USDC (DEX swap ix — fixed at build time).
    pub hop1: Instruction,
    /// Hop 2 template: USDC → B exact-in; standard SPL `Transfer` amount patched from Frame.
    pub hop2_template: Instruction,
    /// Optional hop 2 deliver leg (e.g. pool → user token B); static CPI after patched hop 2.
    pub hop2_deliver: Option<Instruction>,
}

pub fn plan_two_hop_token_swap_instructions(
    scratch: &mut FrameScratch,
    accounts: &TwoHopTokenSwapAccounts,
    hops: &TwoHopTokenSwapInstructions,
) -> Result<Vec<Instruction>, ScratchError> {
    let mut out = Vec::new();
    out.push(scratch.ix_reset());
    out.push(hops.hop1.clone());

    let mut let_batch = scratch.let_builder();
    let usdc_out = let_batch.spl_token_amount(accounts.user_usdc_ata)?;
    out.push(let_batch.build_ix()?);

    let hop2 = build_structured_cpi(
        &hops.hop2_template,
        structured_token_transfer(frame_value(&usdc_out)),
    )?;
    out.push(scratch.ix_cpi(&hop2)?);

    if let Some(deliver) = &hops.hop2_deliver {
        out.push(deliver.clone());
    }

    Ok(out)
}
